File: question4.cpp
/*
 * Name: Travis
 * Student ID: 1113545
 * Date of Submission: 11/28/2024
 *
 * Program to schedule tasks with deadlines using a Max Priority Queue.
 * Features:
 * 1. Prioritize tasks with higher profit.
 * 2. Ensure deadlines are respected while maximizing profit.
 */

#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

// Structure to represent a task
struct Task {
    int profit;   // Profit of the task
    int deadline; // Deadline of the task

    // Constructor
    Task(int p, int d) : profit(p), deadline(d) {}
};

// Comparator for sorting tasks by profit in descending order
bool compareTasks(const Task& t1, const Task& t2) {
    return t1.profit > t2.profit;
}

// Function to schedule tasks to maximize profit
pair<int, vector<int>> scheduleTasks(vector<Task>& tasks, int n) {
    // Step 1: Sort tasks by profit in descending order
    sort(tasks.begin(), tasks.end(), compareTasks);

    // Step 2: Initialize variables
    int maxDeadline = 0;
    for (const auto& task : tasks) {
        maxDeadline = max(maxDeadline, task.deadline);
    }

    vector<int> timeSlots(maxDeadline + 1, -1); // Slots for tasks (-1 indicates free slot)
    vector<int> scheduledTasks; // To store profits of scheduled tasks
    int totalProfit = 0;

    // Step 3: Schedule tasks
    for (const auto& task : tasks) {
        // Try to schedule the task in the latest available slot before its deadline
        for (int slot = task.deadline; slot > 0; --slot) {
            if (timeSlots[slot] == -1) {
                timeSlots[slot] = task.profit; // Assign the task to this slot
                scheduledTasks.push_back(task.profit); // Record the profit
                totalProfit += task.profit; // Add the profit to the total
                break;
            }
        }
    }

    return {totalProfit, scheduledTasks};
}

int main() {
    // Input: Number of tasks
    int n;
    cout << "Enter the number of tasks: ";
    cin >> n;

    vector<Task> tasks;

    // Input: Profit and deadline for each task
    cout << "Enter profit and deadline for each task (separated by a space):" << endl;
    for (int i = 0; i < n; i++) {
        int profit, deadline;
        cin >> profit >> deadline;
        tasks.emplace_back(profit, deadline);
    }

    // Schedule tasks and calculate maximum profit
    auto result = scheduleTasks(tasks, n);

    // Output the results
    cout << "Maximum Profit: " << result.first << endl;
    cout << "Scheduled Tasks: [";
    for (size_t i = 0; i < result.second.size(); i++) {
        cout << result.second[i];
        if (i < result.second.size() - 1) cout << ", ";
    }
    cout << "]" << endl;

    return 0;
}

File: question4.py
/*
 * Name: Travis
 * Student ID: 1113545
 * Date of Submission: 11/28/2024
 *
 * Program to schedule tasks with deadlines using a Max Priority Queue.
 * Features:
 * 1. Prioritize tasks with higher profit.
 * 2. Ensure deadlines are respected while maximizing profit.
 */

import heapq

def schedule_tasks_with_deadlines(tasks):
    # Sort tasks by profit in descending order
    tasks.sort(key=lambda x: x[0], reverse=True)

    # Find the maximum deadline
    max_deadline = max(task[1] for task in tasks)

    # Initialize a result array to track scheduled tasks
    result = [None] * max_deadline

    total_profit = 0

    for profit, deadline in tasks:
        # Try to schedule the task in the latest available slot before or on its deadline
        for i in range(min(deadline, max_deadline) - 1, -1, -1):
            if result[i] is None:
                result[i] = profit
                total_profit += profit
                break

    # Filter out None values to get the list of scheduled tasks
    scheduled_tasks = [task for task in result if task is not None]

    return total_profit, scheduled_tasks

# Input handling
n = int(input("Enter the number of tasks: "))
tasks = []

for _ in range(n):
    profit, deadline = map(int, input("Enter profit and deadline: ").split())
    tasks.append((profit, deadline))

# Schedule tasks
max_profit, scheduled_tasks = schedule_tasks_with_deadlines(tasks)

# Output results
print("Maximum Profit:", max_profit)
print("Scheduled Tasks:", scheduled_tasks)
